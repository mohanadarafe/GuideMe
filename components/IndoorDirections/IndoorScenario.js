import React from 'react';
import { G } from 'react-native-svg';
import { HallXCoordinates } from '../../constants/HallXCoordinates';
import { DifferentFloorDirections } from './TypesOfDirections/DifferentFloorDirections';
import { SameFloorDirections } from './TypesOfDirections/SameFloorDirections';
import { DifferentBuildingDirections } from './TypesOfDirections/DifferentBuildingDirections';
import { getFloorNumber } from './Dijkstra/DijkstraAlgorithm';

export function IndoorScenario(props) {
    const rooms = HallXCoordinates();

    if (props.from && props.to) {
        const routeToTake = whichPathToTake(props.from.toString(), props.to.toString());
        switch (routeToTake) {
            case "SAME_FLOOR":
                return <SameFloorDirections rooms={rooms} floor={props.floor} from={props.from.toString()} to={props.to.toString()}/>
                break;
            case "DIFFERENT_FLOOR":
                return <DifferentFloorDirections rooms={rooms} floor={props.floor} from={props.from.toString()} to={props.to.toString()}/>;
                break;
            case "DIFFERENT_BUILDING":
                return <DifferentBuildingDirections rooms={rooms} floor={props.floor} from={props.from.toString()} to={props.to.toString()}/>;
                break;
            case "NOT_INDOOR":
                break;
        }
    }
    return(
        <G></G>
    );
}

/**
 * Takes two classrooms & determines if the user needs to leave the
 * building, switch floors or stay in the same floor.
 * @param {*} from 
 * @param {*} to 
 */
export const whichPathToTake = (from, to) => {
    if (from && to) {
        const different_building = "DIFFERENT_BUILDING";
        const same_floor = "SAME_FLOOR";
        const different_floor = "DIFFERENT_FLOOR";
        const not_indoor = "NOT_INDOOR";

        if (from.includes(" ") && to.includes(" ")) {
            return not_indoor;
        }
    
        if (from.charAt(0) !== to.charAt(0) && from.length != to.length) {
            return different_building;
        }
    
        if (getFloorNumber(from) == getFloorNumber(to)) {
            return same_floor;
        } else {
            return different_floor;
        }
    }
}
