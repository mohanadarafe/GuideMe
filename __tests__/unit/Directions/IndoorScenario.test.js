import React from "react";
import { IndoorScenario, whichPathToTake } from "../../../components/IndoorDirections/IndoorScenario";
import renderer from "react-test-renderer";

describe("IndoorScenario component", () => {
    test("renders same floor correctly", () => {
        const tree = renderer.create(<IndoorScenario from={"H825"} to={"H801"} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("renders different floor correctly", () => {
        const tree = renderer.create(<IndoorScenario from={"H825"} to={"H501"} />).toJSON();
        expect(tree).toMatchSnapshot();
    })

    test("renders different building correctly", () => {
        const tree = renderer.create(<IndoorScenario from={"H825"} to={"Grey Nuns Annex"} />).toJSON();
        expect(tree).toMatchSnapshot();
    })

    test("returns correct algorithm path", () => {
        const testCases = [
            {from: "H820", to: "H720"},
            {from: "H1025", to: "H155"},
            {from: "H562", to: "H513"},
            {from: "H562", to: "MB3.124"},
            {from: "H1025", to: "H1157"},
            {from: "H961_7", to: "H906"},
            {from: "H961_7", to: "H801"},
            {from: "H961_7", to: "Grey Nuns Annex"},
            {from: "EN Annex", to: "Grey Nuns Annex"},
        ]
        const results = [
            "DIFFERENT_FLOOR",
            "DIFFERENT_FLOOR",
            "SAME_FLOOR",
            "DIFFERENT_BUILDING",
            "DIFFERENT_FLOOR",
            "SAME_FLOOR",
            "DIFFERENT_FLOOR",
            "DIFFERENT_BUILDING",
            "NOT_INDOOR"
        ]
        testCases.forEach((element, index) => {
            expect(whichPathToTake(element.from, element.to)).toEqual(results[index])
        });
    })
});
