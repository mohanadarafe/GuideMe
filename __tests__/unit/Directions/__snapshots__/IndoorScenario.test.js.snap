// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndoorScenario component renders different building correctly 1`] = `
<RNSVGGroup
  fill={
    Array [
      0,
      4278190080,
    ]
  }
  fillOpacity={1}
  fillRule={1}
  font={Object {}}
  matrix={
    Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ]
  }
  opacity={1}
  originX={0}
  originY={0}
  propList={Array []}
  rotation={0}
  scaleX={1}
  scaleY={1}
  skewX={0}
  skewY={0}
  stroke={null}
  strokeDasharray={null}
  strokeDashoffset={null}
  strokeLinecap={0}
  strokeLinejoin={0}
  strokeMiterlimit={4}
  strokeOpacity={1}
  strokeWidth={1}
  vectorEffect={0}
  x={0}
  y={0}
/>
`;

exports[`IndoorScenario component renders different floor correctly 1`] = `
<RNSVGGroup
  fill={
    Array [
      0,
      4278190080,
    ]
  }
  fillOpacity={1}
  fillRule={1}
  font={Object {}}
  matrix={
    Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ]
  }
  opacity={1}
  originX={0}
  originY={0}
  propList={Array []}
  rotation={0}
  scaleX={1}
  scaleY={1}
  skewX={0}
  skewY={0}
  stroke={null}
  strokeDasharray={null}
  strokeDashoffset={null}
  strokeLinecap={0}
  strokeLinejoin={0}
  strokeMiterlimit={4}
  strokeOpacity={1}
  strokeWidth={1}
  vectorEffect={0}
  x={0}
  y={0}
/>
`;

exports[`IndoorScenario component renders same floor correctly 1`] = `
<RNSVGGroup
  fill={
    Array [
      0,
      4278190080,
    ]
  }
  fillOpacity={1}
  fillRule={1}
  font={Object {}}
  matrix={
    Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ]
  }
  opacity={1}
  originX={0}
  originY={0}
  propList={Array []}
  rotation={0}
  scaleX={1}
  scaleY={1}
  skewX={0}
  skewY={0}
  stroke={null}
  strokeDasharray={null}
  strokeDashoffset={null}
  strokeLinecap={0}
  strokeLinejoin={0}
  strokeMiterlimit={4}
  strokeOpacity={1}
  strokeWidth={1}
  vectorEffect={0}
  x={0}
  y={0}
/>
`;
